src/test/scala/com/microsoft/hyperspace/index/rules/JoinIndexRuleV2Test.scala

```diff
@@ -0,0 +1,247 @@
+    t3ScanNode = LogicalRelation(t3Relation, Seq(t3c1, t3c2, t3c3, t3c4), None, false)
+
+    t1FilterNode = Filter(IsNotNull(t1c1), t1ScanNode)
+    t2FilterNode = Filter(IsNotNull(t2c1), t2ScanNode)
+    t3FilterNode = Filter(IsNotNull(t3c1), t3ScanNode)
+
+    t1ProjectNode = Project(Seq(t1c1, t1c3), t1FilterNode)
+    // Project [t1c1#0, t1c3#2]
+    //  +- Filter isnotnull(t1c1#0)
+    //   +- Relation[t1c1#0,t1c2#1,t1c3#2,t1c4#3] parquet
+
+    t2ProjectNode = Project(Seq(t2c1, t2c3), t2FilterNode)
+    // Project [t2c1#4, t2c3#6]
+    //  +- Filter isnotnull(t2c1#4)
+    //   +- Relation[t2c1#4,t2c2#5,t2c3#6,t2c4#7] parquet
+
+    t3ProjectNode = Project(Seq(t3c1, t3c3), t3FilterNode)
+    // Project [t3c1#4, t3c3#6]
+    //  +- Filter isnotnull(t3c1#4)
+    //   +- Relation[t3c1#4,t3c2#5,t3c3#6,t3c4#7] parquet
+
+    createIndexLogEntry("t1i1", Seq(t1c1), Seq(t1c3), t1ProjectNode)
+    createIndexLogEntry("t1i2", Seq(t1c1, t1c2), Seq(t1c3), t1ProjectNode)
+    createIndexLogEntry("t1i3", Seq(t1c2), Seq(t1c3), t1ProjectNode)
+    createIndexLogEntry("t2i1", Seq(t2c1), Seq(t2c3), t2ProjectNode)
+    createIndexLogEntry("t2i2", Seq(t2c1, t2c2), Seq(t2c3), t2ProjectNode)
+    createIndexLogEntry("t3i1", Seq(t3c1), Seq(t3c3), t3ProjectNode)
+  }
+
+  before {
+    spark.conf.set(IndexConstants.INDEX_SYSTEM_PATH, systemPath.toUri.toString)
+    clearCache()
+  }
+
+  private def baseRelation(
+      location: FileIndex,
+      schema: StructType,
+      spark: SparkSession): HadoopFsRelation = {
+    HadoopFsRelation(location, new StructType(), schema, None, new ParquetFileFormat, Map.empty)(
+      spark)
+  }
+
+  override def afterAll(): Unit = {
+    FileUtils.delete(systemPath)
+    super.afterAll()
+  }
+
+  test("Join rule does not update plan for Broadcast Hash Join compatible joins") {
+    withSQLConf("spark.hyperspace.rule.joinV2.enabled" -> "true") {
+      val joinCondition = EqualTo(t1c1, t3c1)
+      val originalPlan =
+        Join(t1ProjectNode, t3ProjectNode, JoinType("inner"), Some(joinCondition))
+      val updatedPlan = JoinIndexRuleV2(originalPlan)
+      assert(updatedPlan.equals(originalPlan))
+    }
+  }
+
+  test("Join rule works if indexes exist and configs are set correctly") {
+    withSQLConf("spark.hyperspace.rule.joinV2.enabled" -> "true") {
+      val joinCondition = EqualTo(t1c1, t2c1)
+      val originalPlan =
+        Join(t1ProjectNode, t2ProjectNode, JoinType("inner"), Some(joinCondition))
+      val updatedPlan = JoinIndexRuleV2(originalPlan)
+      assert(!updatedPlan.equals(originalPlan))
+
+      val indexPaths = getIndexDataFilesPaths("t1i1") ++ getIndexDataFilesPaths("t2i1")
+      verifyUpdatedIndex(originalPlan, updatedPlan, indexPaths)
+    }
+  }
+
+  test("Join rule updates sort merge join part of a plan with both smj and bhj.") {
+    withSQLConf("spark.hyperspace.rule.joinV2.enabled" -> "true") {
+      val bhjCondition = EqualTo(t2c1, t3c1)
```

Comment:
Can you add a small diagram to understand better?
```
      SMJ
      /   \
  BHJ       \
  /  \      T3
T1    T2
```
And point out what is being changed?

Reply:
thanks @imback82 , fixed